### PR TITLE
Move project repo path parsing logic into integration

### PR DESCRIPTION
**Pull request by jianyuan**
This PR moves the project repository path parsing logic into its corresponding integration class. This change will allow other integrations such as Bitbucket and VSTS to implement their own path-parsing logic.

This test should cover the cases: https://github.com/getsentry/sentry/blob/master/tests/sentry/api/endpoints/test_project_repo_path_parsing.py

### CODE REVIEW

File: src/sentry/integrations/github/integration.py [hunk 1]
```diff
@@ -177,11 +177,24 @@ def get_repositories(
     def search_issues(self, query: str) -> Mapping[str, Sequence[Mapping[str, Any]]]:
         return self.get_client().search_issues(query)
 
+    def source_url_matches(self, url: str) -> bool:
```
Comment (AniketDas-Tekky) [inline src/sentry/integrations/github/integration.py:180]: For all these functions,  can you add test cases?

Comment (jianyuan) [reply src/sentry/integrations/github/integration.py:180]: @AniketDas-Tekky Done! I've added some test cases.

**Review by leedongwei — APPROVED** [review]
LGTM 🎉

File: tests/sentry/integrations/github/test_integration.py [hunk 2]
```diff
@@ -880,3 +880,58 @@ def test_get_trees_for_org_falls_back_to_cache_once_MAX_CONNECTION_ERRORS_is_hit
...
+        integration = Integration.objects.get(provider=self.provider.key)
+        repo = Repository.objects.create(
+            organization_id=self.organization.id,
+            name="Test-Organization/repo",
+            url="https://github.com/Test-Organization/repo",
+            provider="integrations:github",
+            external_id=123,
+            config={"name": "Test-Organization/repo"},
+            integration_id=integration.id,
+        )
```
Comment (leedongwei) [inline tests/sentry/integrations/github/test_integration.py:914]: Can you wrap this with `with assume_test_silo_mode(SiloMode.REGION):` (see [example here](https://github.com/getsentry/sentry/blob/master/tests/sentry/integrations/github/test_integration.py#L230-L238)). This should resolve the failing test.

Comment (leedongwei) [reply tests/sentry/integrations/github/test_integration.py:914]: https://github.com/getsentry/sentry/blob/master/tests/sentry/integrations/github/test_integration.py#L230-L238

Comment (leedongwei) [reply tests/sentry/integrations/github/test_integration.py:914]: For context, we're splitting up our codebase + database to deploy in separate silos so we can differentiate between Sentry user data vs Sentry users' user data.

As such, some models are not available everywhere and this wrapper helps the tests pretend that they were created in the correct silo.

Comment (jianyuan) [reply tests/sentry/integrations/github/test_integration.py]: @leedongwei Done!

**Review by leeandher — APPROVED** [review]

**Pull request merged**